src/vm_pages.rs

```diff
@@ -1751,6 +1752,100 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
         Ok(())
     }
 
+    /// Invalidates a page range.
+    pub fn block_pages(&self, page_addr: GuestPageAddr, len: u64) -> Result<()> {
+        let invalidated = self
+            .inner
+            .root
+            .invalidate_range_sparse(page_addr, len, |addr| {
+                self.inner
+                    .page_tracker
+                    .is_mapped_page(addr, self.inner.page_owner_id, MemType::Ram)
+                    || (self.inner.page_tracker.is_shared_page(addr, MemType::Ram)
+                        && self.inner.page_tracker.is_owned(addr, PageOwnerId::host()))
+            })
+            .map_err(Error::Paging)?;
+        let version = self.inner.tlb_tracker.current_version();
+        for paddr in invalidated {
+            // Safety: We've verified the typing of the page and its ownership
+            // before it was invalidated.
+            let page: Page<Invalidated> = unsafe { Page::new(paddr) };
+            // Unwrap ok: Page was mapped or shared and has just been invalidated.
+            self.inner.page_tracker.block_page(page, version).unwrap();
+        }
+
+        Ok(())
+    }
+
+    /// Unblocks previously invalidated page range.
+    pub fn unblock_pages(&self, page_addr: GuestPageAddr, len: u64) -> Result<()> {
+        let valid_pages = self
+            .inner
+            .root
+            .validate_range(page_addr, len, |addr| {
+                self.inner.page_tracker.is_blocked_page(
+                    addr,
+                    self.inner.page_owner_id,
+                    MemType::Ram,
+                ) || self
+                    .inner
+                    .page_tracker
+                    .is_blocked_shared_page(addr, MemType::Ram)
+            })
```

Comment:
No that's the only place where we use it, so I can merge this into a single helper. And I think this will be the same for `promote`/`demote` functions, all they'll care about is if the page is blocked, doesn't matter if it really is `BlockedShared` or `Blocked`.